content/posts/2023-11-20-tools-for-learning-llvm-tablegen.md

```diff
@@ -0,0 +1,433 @@
+---
+author: "David Spickett"
+date: "2023-11-20"
+tags: ["TableGen"]
+title: "Tools for Learning LLVM TableGen"
+---
+
+[TableGen](https://github.com/llvm/llvm-project/tree/main/llvm/utils/TableGen)
+is a language used within the LLVM project for generating all kinds of files,
+when maintaining those files by hand would be very difficult.
+
+For example, it is used to define all of the instructions that can be used on a
+particular architecture. The information is defined once in TableGen and we can
+produce many things based on it. C++ code, documentation, command line options,
+etc.
+
+TableGen has been around since
+[before](https://github.com/llvm/llvm-project/commit/a6240f6b1a34f9238cbe8bc8c9b6376257236b0a)
+the first official release of LLVM, over 20 years ago.
+
+Today in the monorepo there are over a thousand TableGen source files totalling
+over 500,000 lines of code. Making it the 5th most popular language in the
+monorepo.
+
+| Language     | files | blank  | comment   | code     |
+| ------------ | ------| -------| --------- | -------- |
+| C++          | 29642 | 958542 |  1870101  | 5544445  |
+| C/C++ Header | 11844 | 316806 |   499845  | 1486165  |
+| C            | 10535 | 259900 |  1603594  | 1011269  |
+| Assembly     | 10694 | 478035 |  1222315  |  820236  |
+| TableGen     |  1312 |  94112 |    83616  |  580289  |
+
+(Counted from
+[this commit](https://github.com/llvm/llvm-project/commit/ba24b814f2a20a136f0a7a0b492b6ad8a62114c6),
+rest of table omitted)
+
+With projects like MLIR
+[embracing TableGen](https://mlir.llvm.org/docs/DefiningDialects/Operations/)
+it’s only going to grow from here. So if you’re contributing to LLVM, you’re
+going to encounter it eventually.
+
+Which might be a problem, because TableGen only exists within LLVM. It doesn’t
+have the massive array of resources that a language like C++ has.
+
+On top of joining a new project you have this Domain Specific Language (DSL)
+to learn. You didn’t come to LLVM to learn a DSL, you probably came here
+to write a compiler.
+
+I can’t say that this problem will ever be solved, but the situation isn’t as
+bleak as it appears. There have been big improvements in TableGen tools
+recently, which means you can put more of your energy into the goals that
+brought you to LLVM in the first place.
+
+# A Brief Introduction to TableGen
+
+Imagine you wanted to represent the registers of an architecture. I’m going to
+use Arm's AArch64 in particular here.
+
+You could describe them in TableGen like this:
+```
+class Register<int _size, string _alias=""> {
+   int size = _size;
+   string alias = _alias;
+}
+
+// 64 bit general purpose registers are X<N>.
+def X0: Register<8> {}
+// Some have special alternate names.
+def X29: Register<8, "frame pointer"> {}
+// ...a lot more registers omitted.
+```
+
+By default the TableGen compiler `llvm-tblgen` produces what are called
+“records”, which are shown below.
```

Comment:
By default, the TabelGen compiler llvm-tblgen creates "records" - which are shown below.

Reply:
Done.